File: src/lib/keychain.ts
/**
 * Typed wrapper around the backend keychain Tauri commands.
 *
 * The backend stores credentials in the macOS Keychain (on macOS) or no-ops
 * with a clear error on other platforms. Call {@link keychainAvailable} once
 * at UI mount to gate "Save to Keychain" controls; it hides the feature
 * gracefully on unsupported platforms.
 */

import { invoke } from '@tauri-apps/api/core';

export type CredentialKind =
  | 'ssh_password'
  | 'ssh_key_passphrase'
  | 'sftp_password'
  | 'sftp_key_passphrase'
  | 'ftp_password';

/**
 * Whether the backend can actually read/write the OS keychain on this build.
 * Returns false for any invoke failure so the UI falls back to a plain
 * password prompt even if the command isn't registered for some reason.
 */
export async function keychainAvailable(): Promise<boolean> {
  try {
    return await invoke<boolean>('keychain_available');
  } catch {
    return false;
  }
}

export async function keychainSave(
  kind: CredentialKind,
  account: string,
  secret: string,
): Promise<void> {
  await invoke('keychain_save', { kind, account, secret });
}

export async function keychainLoad(
  kind: CredentialKind,
  account: string,
): Promise<string | null> {
  const result = await invoke<string | null>('keychain_load', {
    kind,
    account,
  });
  return result ?? null;
}

export async function keychainDelete(
  kind: CredentialKind,
  account: string,
): Promise<void> {
  await invoke('keychain_delete', { kind, account });
}

/**
 * Stable Keychain account identifier for a (host, port, username) tuple.
 * Using this consistently means the same credential is reused across
 * reconnects to the same endpoint, and distinct endpoints never collide.
 */
export function accountFor(
  host: string,
  port: number,
  username: string,
): string {
  return `${username}@${host}:${port}`;
}

/**
 * Map a (protocol, authMethod) pair to the CredentialKind that identifies
 * its stored secret, or null if this combination has no secret to store
 * (e.g. anonymous FTP, or public-key auth without a passphrase).
 */
export function credentialKindFor(
  protocol: string,
  authMethod: string,
): CredentialKind | null {
  const p = protocol.toUpperCase();
  if (p === 'SSH') {
    if (authMethod === 'password') return 'ssh_password';
    if (authMethod === 'publickey') return 'ssh_key_passphrase';
  } else if (p === 'SFTP') {
    if (authMethod === 'password') return 'sftp_password';
    if (authMethod === 'publickey') return 'sftp_key_passphrase';
  } else if (p === 'FTP') {
    if (authMethod === 'password') return 'ftp_password';
  }
  return null;
}

File: src-tauri/src/commands/keychain.rs
//! Tauri commands exposing the macOS Keychain for SSH / SFTP / FTP credentials.
//!
//! The frontend flow is:
//! 1. `keychain_available()` — call once at startup. When false, hide any
//!    "Save to Keychain" UI and fall back to plain-password prompts.
//! 2. On connect form load: `keychain_load(kind, account)` to prefill the
//!    password if a saved entry exists.
//! 3. After a successful connect with the "save" checkbox: `keychain_save(
//!    kind, account, secret)`.
//! 4. On user-initiated credential removal: `keychain_delete(kind, account)`.
//!
//! `account` is an opaque string chosen by the frontend; the convention is
//! `"<username>@<host>:<port>"` so the same credential is reused across
//! reconnects to the same endpoint.

use crate::keychain::{self, CredentialKind};

#[tauri::command]
pub async fn keychain_available() -> Result<bool, String> {
    Ok(keychain::is_supported())
}

#[tauri::command]
pub async fn keychain_save(
    kind: CredentialKind,
    account: String,
    secret: String,
) -> Result<(), String> {
    keychain::save_password(kind, &account, &secret).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn keychain_load(
    kind: CredentialKind,
    account: String,
) -> Result<Option<String>, String> {
    keychain::load_password(kind, &account).map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn keychain_delete(kind: CredentialKind, account: String) -> Result<(), String> {
    keychain::delete_password(kind, &account).map_err(|e| e.to_string())
}

/// List all accounts stored under the given credential kind. Returns an empty
/// array — never an error — when no entries exist or the platform has no
/// keychain, so the UI can render a "nothing saved" state unconditionally.
#[tauri::command]
pub async fn keychain_list(kind: CredentialKind) -> Result<Vec<String>, String> {
    keychain::list_accounts(kind).map_err(|e| e.to_string())
}
